actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,459 @@
+		return nil, err
+	}
+	if len(stateList) != len(metastateList) {
+		return nil, errs.Errorf("inconsistent number of states and metatstates in the current work item type (must be equal): %d states != %d metastates", len(stateList), len(metastateList))
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errs.New("metastate value in value list is not of type string")
+		}
+		// only the first state<->metatstate mapping needs to be taken as
+		// per definition. So we're only adding a new entry here if there
+		// is not an entry already existing.
+		if _, ok := stateToMetastateMap[thisState]; !ok {
+			stateToMetastateMap[thisState] = thisMetastate
+		}
+	}
+	return stateToMetastateMap, nil
+}
+
+// getMetastateToStateMap returns the mapping from metastate to state values read from the template.
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
+	for _, change := range changes {
+		if change.AttributeName == attributeName {
+			change.NewValue = newValue
+			return changes
+		}
+	}
+	newChanges := append(changes, change.Change{
+		AttributeName: attributeName,
+		OldValue:      oldValue,
+		NewValue:      newValue,
+	})
+	return newChanges
+}
+
+func (act ActionStateToMetaState) fuseChanges(c1 change.Set, c2 change.Set) change.Set {
+	for _, change := range c2 {
+		c1 = act.addOrUpdateChange(c1, change.AttributeName, change.OldValue, change.NewValue)
+	}
+	return c1
+}
+
+// OnChange executes the action rule. It implements rules.Action. Returns the Work Item with eventual changes applied.
+func (act ActionStateToMetaState) OnChange(newContext change.Detector, contextChanges change.Set, configuration string, actionChanges change.Set) (change.Detector, change.Set, error) {
+	if act.Ctx == nil {
+		return nil, nil, errs.New("context is nil")
+	}
+	if act.Db == nil {
+		return nil, nil, errs.New("database is nil")
+	}
+	if act.UserID == nil {
+		return nil, nil, errs.New("userID is nil")
+	}
+	if actionChanges == nil {
```

Comment:
@kwk Is this the correct way to check for empty `actionChanges`? Earlier it was pointer and this check made sense but now it represents `[]changes` and each change is
```go
type Set []Change

type Change struct {
	AttributeName string
	NewValue      interface{}
	OldValue      interface{}
}
```

Should this line be replaced with `len(actionChanges) > 0` ?

Reply:
The whole check does not make sense for the non-pointer scenario. This was in there to check that there is actually a usable `actionChanges` given. As we now always have an instance, this check is redundant. The `actionChanges` is allowed to be empty here.

Reply:
89c8cb4